File: clients/JavaScript/model/debug-command.js
const DebugData = require('./debug-data');
class DebugCommand {
    static async readFrom(stream) {
        let tag = await stream.readInt();
        if (tag == Add.TAG) {
            return await Add.readFrom(stream);
        }
        if (tag == Clear.TAG) {
            return await Clear.readFrom(stream);
        }
        if (tag == SetAutoFlush.TAG) {
            return await SetAutoFlush.readFrom(stream);
        }
        if (tag == Flush.TAG) {
            return await Flush.readFrom(stream);
        }
        throw new Error("Unexpected tag value");
    }
}

class Add extends DebugCommand {
    constructor(data) {
        super();
        this.data = data;
    }
    static async readFrom(stream) {
        let data;
        data = await DebugData.readFrom(stream);
        return new Add(data);
    }
    async writeTo(stream) {
        await stream.writeInt(Add.TAG);
        let data = this.data;
        await data.writeTo(stream);
    }
}
DebugCommand.Add = Add;
Add.TAG = 0;
class Clear extends DebugCommand {
    constructor() {
        super();
    }
    static async readFrom(stream) {
        return new Clear();
    }
    async writeTo(stream) {
        await stream.writeInt(Clear.TAG);
    }
}
DebugCommand.Clear = Clear;
Clear.TAG = 1;
class SetAutoFlush extends DebugCommand {
    constructor(enable) {
        super();
        this.enable = enable;
    }
    static async readFrom(stream) {
        let enable;
        enable = await stream.readBool();
        return new SetAutoFlush(enable);
    }
    async writeTo(stream) {
        await stream.writeInt(SetAutoFlush.TAG);
        let enable = this.enable;
        await stream.writeBool(enable);
    }
}
DebugCommand.SetAutoFlush = SetAutoFlush;
SetAutoFlush.TAG = 2;
class Flush extends DebugCommand {
    constructor() {
        super();
    }
    static async readFrom(stream) {
        return new Flush();
    }
    async writeTo(stream) {
        await stream.writeInt(Flush.TAG);
    }
}
DebugCommand.Flush = Flush;
Flush.TAG = 3;
module.exports = DebugCommand;

File: clients/Ruby/model/debug_command.rb
require_relative 'debug_data'
class DebugCommand
    def self.read_from(stream)
        tag = stream.read_int()
        if tag == DebugCommand::Add::TAG
            return DebugCommand::Add.read_from(stream)
        end
        if tag == DebugCommand::Clear::TAG
            return DebugCommand::Clear.read_from(stream)
        end
        if tag == DebugCommand::SetAutoFlush::TAG
            return DebugCommand::SetAutoFlush.read_from(stream)
        end
        if tag == DebugCommand::Flush::TAG
            return DebugCommand::Flush.read_from(stream)
        end
        raise "Unexpected tag value"
    end

    class Add
        TAG = 0
        attr_accessor :data
        def initialize(data)
            @data = data
        end
        def self.read_from(stream)
            data = DebugData.read_from(stream)
            Add.new(data)
        end
        def write_to(stream)
            stream.write_int(TAG)
            @data.write_to(stream)
        end
    end
    class Clear
        TAG = 1
        def initialize()
        end
        def self.read_from(stream)
            Clear.new()
        end
        def write_to(stream)
            stream.write_int(TAG)
        end
    end
    class SetAutoFlush
        TAG = 2
        attr_accessor :enable
        def initialize(enable)
            @enable = enable
        end
        def self.read_from(stream)
            enable = stream.read_bool()
            SetAutoFlush.new(enable)
        end
        def write_to(stream)
            stream.write_int(TAG)
            stream.write_bool(@enable)
        end
    end
    class Flush
        TAG = 3
        def initialize()
        end
        def self.read_from(stream)
            Flush.new()
        end
        def write_to(stream)
            stream.write_int(TAG)
        end
    end
end
